### PR TITLE
(CDAP-9314) Fix Kryo serialization problem for StructuredRecord

### PR DESCRIPTION
**Pull request by chtyim**
The changes are already group into commits. This change is larger than needed for CDAP-9314 itself, however it unified how to serialize and deserialize StructureRecord and is beneficial to other bug fixes/improvements (e.g [CDAP-8984](https://issues.cask.co/browse/CDAP-8984)).

Here are the explanations for each commits:

1. Created the Kryo Serializer for Schema class
   * Add it as the default serializer by rewriting the Kryo class constructor. 
   * This essentially is enough to fix CDAP-9314.
2. Move Encoder, Decoder, DatumReader and DatumWriter interfaces to cdap-formats.
   * Just refactoring to prepare for having implementations for StructuredRecord in cdap-formats module.
3. Create StructuredRecord datum reader and writer
   * Introduced the base code for serializing and deserializing `StructureRecord`.
   * Also introduced the Json version of it and refactor existing util class `StructuredRecordStringConverter` to use them
   * If there are other places that involves serde of StructuredRecord, we can use these classes, combined with `BinaryEncoder` and `BinaryDecoder` to be more efficient
4. Created the Kryo Serializer for StructuredRecord
   * This is quite trivial, no need to explain.

### CODE REVIEW

**Comment by chtyim** [comment]
Build link https://builds.cask.co/browse/CDAP-RUT886

File: cdap-formats/src/main/java/co/cask/cdap/format/io/JsonStructuredRecordDatumReader.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    jsonReader.endArray();
+
+    return array;
+  }
+
+  @Override
+  protected Map<?, ?> decodeMap(Decoder decoder, Schema keySchema, Schema valueSchema) throws IOException {
+    if (!keySchema.isCompatible(Schema.of(Schema.Type.STRING))) {
+      throw new IOException("Complex key type not supported: " + keySchema);
```
Comment (albertshau) [inline cdap-formats/src/main/java/co/cask/cdap/format/io/JsonStructuredRecordDatumReader.java]: may not be immediately obvious to users what 'key' refers to, a little better to say 'Complex key type in maps are not supported: '

File: cdap-formats/src/main/java/co/cask/cdap/format/io/JsonStructuredRecordDatumReader.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    JsonReader jsonReader = getJsonReader(decoder);
+    JsonToken token = jsonReader.peek();
+    // Based on the token to guess the schema
+    for (Schema schema : unionSchema.getUnionSchemas()) {
+      if (SCHEMA_TO_JSON_TYPE.get(schema.getType()) == token) {
+        return decode(decoder, schema);
+      }
+    }
+
+    throw new IOException("No matching schema found for union type: " + unionSchema);
```
Comment (albertshau) [inline cdap-formats/src/main/java/co/cask/cdap/format/io/JsonStructuredRecordDatumReader.java]: better to include the token that we couldn't find the match for in the error message

**Review by albertshau — COMMENTED** [review]
just a couple comments about exception messages

**Comment by albertshau** [comment]
lgtm

**Comment by chtyim** [comment]
Next time better squash before merging.

**Pull request merged**